core/trino-main/src/main/java/io/trino/operator/TopNRankingOperator.java

```diff
@@ -291,35 +357,59 @@ public Page getOutput()
             unfinishedWork = null;
         }
 
-        if (!finishing) {
+        if (!finishing && (groupedTopNBuilder == null || !groupedTopNBuilder.isFull()) && outputIterator == null) {
             return null;
         }
 
-        if (outputIterator == null) {
+        if (outputIterator == null && groupedTopNBuilder != null) {
             // start flushing
             outputIterator = groupedTopNBuilder.buildResult();
         }
 
         Page output = null;
-        if (outputIterator.hasNext()) {
+        if (outputIterator != null && outputIterator.hasNext()) {
             // rewrite to expected column ordering
             output = outputIterator.next().getColumns(outputChannels);
         }
+        else {
+            closeGroupedTopNBuilder();
+            return null;
+        }
         updateMemoryReservation();
         return output;
     }
 
+    @Override
+    public void close()
+    {
+        closeGroupedTopNBuilder();
+    }
+
+    private void closeGroupedTopNBuilder()
+    {
+        outputIterator = null;
+        groupedTopNBuilder = null;
+        localMemoryContext.setBytes(0);
+    }
+
     @VisibleForTesting
     public int getCapacity()
     {
-        checkState(groupByHash != null);
-        return groupByHash.getCapacity();
+        return groupedTopNBuilder == null ? 0 : groupedTopNBuilder.getCapacity();
     }
 
     private boolean updateMemoryReservation()
     {
+        if (groupedTopNBuilder == null) {
+            localMemoryContext.setBytes(0);
+            return true;
+        }
         // TODO: may need to use trySetMemoryReservation with a compaction to free memory (but that may cause GC pressure)
-        localUserMemoryContext.setBytes(groupedTopNBuilder.getEstimatedSizeInBytes());
+        localMemoryContext.setBytes(groupedTopNBuilder.getEstimatedSizeInBytes());
+        if (groupedTopNBuilder.isFlushable()) {
```

Comment:
This should probably a check on the "partial" boolean flag, rather than the state of the grouped top n builder

Reply:
Done.